{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Find Unique\n",
    " - Deliverables:\n",
    "     - findUnique.py - 50 points ( We will test this code)\n",
    "     - Notebook with Inspection materials and Results ( 5 pts )\n",
    "     - required input from STDIN\n",
    "     - required output to STDOUT\n",
    "     - output must be sorted by tRNA header\n",
    "\n",
    "## Sets and tRNA\n",
    "\n",
    "Mitochondrial tRNA are encoded in the mitochondrial genome and account for ~10% of the coding space, yet over 50% of mitochondrial genomic disease have their origin in this molecule. These molecules are transcribed, processed, modified, and ultimately folded to produce these mature adapters between the mitochondrial transcription and translation processes. Defects in mt.tRNA maturation reflect additional disease states .. if only we could identify those mutations. We are working on such a device, though we first need to use it to identify abundance of these molecules in a mixed population. Are there any unique subsequences among the 22 mt.tRNA that can be used as \"tags\"? If so, we can count those tags to assess abundance.\n",
    "\n",
    "## Assignment\n",
    "\n",
    "Write a python command line program that reads a file of fasta sequences from STDIN, finds the unique subsequences that occur in each single tRNA such that no members of this set occur among any of the other tRNA sets. Each of those 22 sets should be minimized, such that no member of that unique subsequence set is a substring of any other member of this set. [ Unique and Essential]\n",
    "\n",
    "As an example, let's say that both ACG and AAACGA are in a unique set. Since ACG is a substring of AAACGA we would remove AAACGA. [ ACG is Essential ]\n",
    "\n",
    "Use Python sets for this assignment[__required__]. Not only will your code be smaller, but it will be more likely to work. The union, intersection and difference operators will be quite useful.\n",
    "\n",
    "## Rough design plan...\n",
    "0) Consider a tRNA as an object. In thi scase we would then have 22 objects that woul dneed to be able to communicate with each other.\n",
    "\n",
    "1) compute the set of all substrings from each tRNA sequence. [powerset] I will refer to a set of substrings as a tRNAset.\n",
    "\n",
    "2) for each tRNAset, compute the union of all other tRNA sets, and then remove that union from the current tRNAset. Notice that this union operation finds all of the substrings from all other tRNA. If any of those are present in your current tRNA, then they are not unique ! [ Unique]\n",
    "\n",
    "3) for each unique tRNAset, it now contains the truly unique ones, along with any extensions of that subsequence. If, for example, it was found that G only occurred in a single tRNA, then adding an A onto that G must also be unique because it has a G in it. We only want the minimal form.. G. [ Essential]\n",
    "\n",
    "4) Remove spaces from the header line before sorting and printing. This will make your output a little prettier.\n",
    "\n",
    "5) make sure to remove any alignment characters from your initial sequences. These characters are periods(.), underscores(\\_), or dashes(-) . You will find many new characters in the sequence other than {ACGU} - leave these in place, they are modified bases.\n",
    "\n",
    "## Report\n",
    "Print a report that contains items as follows. \n",
    "\n",
    " - Line 1: the tRNA name\n",
    " - Line 2: the tRNA sequence ( with the alignment characters removed)\n",
    " - lines 3-80 or so, each unique element.\n",
    "\n",
    "These unique elements need to be ordered by their starting position in the tRNA sequence, and should be spaced over so they are directly under that subsequence in the original tRNA. This looks like an alignment, but you can find where it belongs by using the string.find() method. Include dots in all positions to the left, serving as alignment guides for your reader. [ see sample output below ]\n",
    "\n",
    "Do this for all of the 22 tRNA sequences.\n",
    "\n",
    "Print the tRNA out as above, sorted by the header line.  \n",
    "\n",
    "## Hints:\n",
    "\n",
    "__use sets and the set operators - this is required !__\n",
    "\n",
    "your final code will be under 100 lines.\n",
    "\n",
    "Do most of your coding using methods defined in a class and then instantiated as objects. \n",
    "\n",
    "The sequences include characters that are just alignment characters. They are not part of the sequence and must be removed. [-\\_\\.] are alignment characters. \n",
    "\n",
    "When removing items from a tRNAset, don't do this while iterating through that set. Also, when building a unique set you will need the original contents of all other tRNAsets. So.. build a new set to keep the unique contents, or keep track of the elements you intend to delete later. Notice that you build the union of all other tRNA, and this happens 22 times - these unions are all distinct from each other. Example, consider 4 sets, A,B,C,D.  we would compute the set B, C, D to use against set A, and we would compute the set A, C, D to use against B. A and B need to not change while we are doing this computation.\n",
    "\n",
    "There are cases where a unique element is present multiple times in a single tRNA. This sequence is unique to this tRNA and should be reported in its multiple locations ( same sequence on multiple lines of the output).\n",
    "\n",
    "This is a command line program, though it does not have any optional parameters. You really don't need the commandline class. Your input comes from STDIN and output goes to STDOUT. Use the FastaReader for input and print statement for output. If you are going to use jupyter to develop your code, you can add a filename to fastareader for your testing.\n",
    "\n",
    "## Submission\n",
    "\n",
    "Submit code using Canvas. As always, work with your group and write your own code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Feedback\n",
    "\n",
    "Mona: Hi Emmet! Yay last push! I like how you made your powerset work. This method was something that I personally struggled with. I think your code was easy to follow along with. I think something for next time could be creating a code that is a little less redundant for example you created individual tRNA 'codes', so maybe you can find something that is able to just refer to the sequence so that you don't have to code it out individually. But this again, just goes to show how thoughtful you were and how good at coding you are haha! Great job: ).\n",
    "\n",
    "Clare: Hi Emmet! Great work. I think the code looks great; the only thing you're missing is comments and docstrings. Also your main could maybe be optimized but that's definitely optional since your code works.\n",
    "\n",
    "Liam: Hi Emmet, great job on your code! I think the modularity on your code and for the TRNA class is great, you've contained the attributes and methods into the class really well. I chose to take a different approach and did most of the logic outside of the tRNA class this time. Great job :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SequenceAnalysis import FastAreader\n",
    "import sys \n",
    "class TRNA():\n",
    "    \"\"\"\n",
    "    Class to represent a tRNA sequence\n",
    "    \"\"\"\n",
    "    list = [] #global list variable will contain all instances of the class \n",
    "    def powerset(self,s):\n",
    "        \"\"\"\n",
    "        Return the powerset of a given string\n",
    "        \"\"\"\n",
    "        result = set()\n",
    "        n = len(s)\n",
    "        for i in range(n):\n",
    "            for j in range(i, n):\n",
    "                result.add(s[i:j+1])\n",
    "        return result\n",
    "    def __init__(self,header,seq):\n",
    "        \"\"\"\n",
    "        Initializes an instance of the class with a name seqeunce, amino acid name, and a powerset and adds itself to the list\n",
    "        \"\"\"\n",
    "        self.name = header.replace(\" \",\"\")\n",
    "        self.aa = self.getAAname()\n",
    "        self.seq = seq\n",
    "        self.powerSet = self.powerset(seq)\n",
    "        self.list.append(self)\n",
    "    def getseq(self):\n",
    "        \"\"\"\n",
    "        return seqeunce \n",
    "        \"\"\"\n",
    "        return self.seq\n",
    "    def uniqueSeq(self):\n",
    "        \"\"\"\n",
    "        Returns essential unique seqeunces\n",
    "        \"\"\"\n",
    "        localList = self.list.copy()\n",
    "        localList.remove(self) #list of all tRNA objects not including itself\n",
    "        set1 = set()\n",
    "        for index,val in enumerate(localList):\n",
    "            set1 = set1.union(val.powerSet) #union of all powersets \n",
    "        set2 = list(self.powerSet - set1) #list of unique substrings\n",
    "        set3 = set2.copy()\n",
    "        for index, val in enumerate(set2): #loop compares each value in the list to each other value and deletes values containing other values until only essentials remain\n",
    "            for i in range(0,len(set2)):\n",
    "                if val != set2[i] and val in set2[i]:\n",
    "                    try:\n",
    "                        set3.remove(set2[i])\n",
    "                    except ValueError:\n",
    "                        continue\n",
    "        self.clearList() #clear list to prevent screw ups down the line \n",
    "        return set3\n",
    "    def clearList(self):\n",
    "        \"\"\"\n",
    "        Clears global list variable\n",
    "        \"\"\"\n",
    "        self.list = []\n",
    "    def find_all_positions(self,string, substring):\n",
    "        \"\"\"\n",
    "        Returns all positions where a given substring is found in a parent string\n",
    "        \"\"\"\n",
    "        start = 0\n",
    "        positions = []\n",
    "        while True:\n",
    "            index = string.find(substring, start)\n",
    "            if index == -1:\n",
    "                break\n",
    "            positions.append(index)\n",
    "            start = index + 1\n",
    "        return positions\n",
    "    def getAAname(self):\n",
    "        \"\"\"\n",
    "        Returns the name of the amino acid associated with a tRNA\n",
    "        \"\"\"\n",
    "        return1 = self.name.split('|')[1]\n",
    "        return return1\n",
    "    def printer(self):\n",
    "        \"\"\"\n",
    "        Prints all essential unique sequences alligned with their position in the sequence\n",
    "        \"\"\"\n",
    "        returnStr = \"\"\n",
    "        out = self.uniqueSeq()\n",
    "        printer = []\n",
    "        sequence = self.getseq()\n",
    "        for item in out: \n",
    "            pos = self.find_all_positions(sequence,item) #find all positions assoicated with an essential \n",
    "            for position in pos:\n",
    "                printer.append([str(item),position])\n",
    "        printer = sorted(printer, key=lambda x: x[1]) #sort by position in the sequence\n",
    "        sys.stdout.reconfigure(encoding='utf-8') #handles special characters\n",
    "        print(self.name)\n",
    "        print(sequence)\n",
    "        for item in printer:\n",
    "            print('.'*item[1]+str(item[0])) #print with number of periods that matches position to allign with the sequence\n",
    "\n",
    "def main(infile=None):\n",
    "    #if infile is None:\n",
    "        #infile = sys.argv[1]\n",
    "    trna_list = []\n",
    "    #with open(infile, 'r', encoding='utf-8') as f:\n",
    "    reader = FastAreader(infile)\n",
    "    for head,seq in reader.readFasta():\n",
    "        #stripping allignment characters and white space\n",
    "        seq = seq.replace(\"-\",\"\")\n",
    "        seq = seq.replace(\"_\",\"\")\n",
    "        seq = seq.replace(\".\",\"\")\n",
    "        trna_list.append([head,seq])\n",
    "    \"\"\"\n",
    "    I deserve whatever I get for this monstrous list of tRNA instantiation but it seemed like the easiest way at the time and now i just feel if it aint broke... \n",
    "    \"\"\"\n",
    "    Trna1 = TRNA(trna_list[0][0], trna_list[0][1])\n",
    "    Trna2 = TRNA(trna_list[1][0], trna_list[1][1])\n",
    "    Trna3 = TRNA(trna_list[2][0], trna_list[2][1])\n",
    "    Trna4 = TRNA(trna_list[3][0], trna_list[3][1])\n",
    "    Trna5 = TRNA(trna_list[4][0], trna_list[4][1])\n",
    "    Trna6 = TRNA(trna_list[5][0], trna_list[5][1])\n",
    "    Trna7 = TRNA(trna_list[6][0], trna_list[6][1])\n",
    "    Trna8 = TRNA(trna_list[7][0], trna_list[7][1])\n",
    "    Trna9 = TRNA(trna_list[8][0], trna_list[8][1])\n",
    "    Trna10 = TRNA(trna_list[9][0], trna_list[9][1])\n",
    "    Trna11 = TRNA(trna_list[10][0], trna_list[10][1])\n",
    "    Trna12 = TRNA(trna_list[11][0], trna_list[11][1])\n",
    "    Trna13 = TRNA(trna_list[12][0], trna_list[12][1])\n",
    "    Trna14 = TRNA(trna_list[13][0], trna_list[13][1])\n",
    "    Trna15 = TRNA(trna_list[14][0], trna_list[14][1])\n",
    "    Trna16 = TRNA(trna_list[15][0], trna_list[15][1])\n",
    "    Trna17 = TRNA(trna_list[16][0], trna_list[16][1])\n",
    "    Trna18 = TRNA(trna_list[17][0], trna_list[17][1])\n",
    "    Trna19 = TRNA(trna_list[18][0], trna_list[18][1])\n",
    "    Trna20 = TRNA(trna_list[19][0], trna_list[19][1])\n",
    "    Trna21 = TRNA(trna_list[20][0], trna_list[20][1])\n",
    "    Trna22 = TRNA(trna_list[21][0], trna_list[21][1])\n",
    "    trnas = [Trna1, Trna2, Trna3, Trna4, Trna5, Trna6, Trna7, Trna8, Trna9, Trna10, Trna11, Trna12, Trna13, Trna14, Trna15, Trna16, Trna17, Trna18, Trna19, Trna20, Trna21, Trna22]\n",
    "    trnas = sorted(trnas, key=lambda x: x.aa) #sorting tRNAs alpha by amino acid \n",
    "    sys.stdout.reconfigure(encoding='utf-8')#handles special characters\n",
    "    for item in trnas: #printing all essentials for each tRNA in alpha order\n",
    "        item.printer()\n",
    "    Trna1.clearList\n",
    "if __name__ == \"__main__\":\n",
    "    sys.stdin.reconfigure(encoding='utf-8')#handles special characters\n",
    "    main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tRNA|Lys|∃UU|Bostaurus|mitochondrial\n",
    "CACUAAGA\"LCUAUAUAGCACPAACCU∃UU6AGUUAGAGAUUGAGAGCCAU\"UACUCUCCUUGGUGACCA\n",
    "CACU\n",
    ".ACUA\n",
    "...UAA\n",
    "....AAG\n",
    ".......A\"\n",
    ".........L\n",
    "..........CUAU\n",
    "............AUAU\n",
    "..............AUAG\n",
    "...............UAGC\n",
    ".................GCA\n",
    ".....................P\n",
    "......................AAC\n",
    ".......................ACCU\n",
    "...........................∃\n",
    "..............................6\n",
    "...............................AGU\n",
    "................................GUU\n",
    ".................................UUA\n",
    "..................................UAGA\n",
    "...................................AGAGA\n",
    "......................................GAU\n",
    ".......................................AUU\n",
    "........................................UUGA\n",
    ".........................................UGAG\n",
    "..........................................GAGAG\n",
    "............................................GAGC\n",
    "..............................................GCC\n",
    "................................................CAU\n",
    "..................................................U\"\n",
    "...................................................\"U\n",
    "....................................................UAC\n",
    ".....................................................ACUC\n",
    ".......................................................UCU\n",
    ".........................................................UCC\n",
    "...........................................................CUU\n",
    "..............................................................GG\n",
    "...............................................................GUG\n",
    ".................................................................GAC\n",
    "..................................................................ACCA"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
